pil-analyzer/src/traits_processor.rs

```diff
@@ -0,0 +1,166 @@
+use core::panic;
+use std::collections::{HashMap, HashSet};
+
+use powdr_ast::{
+    analyzed::{
+        Expression, FunctionValueDefinition, PolynomialReference, Reference, Symbol,
+        TypedExpression,
+    },
+    parsed::{types::Type, FunctionCall, TraitImplementation},
+};
+
+pub struct TraitsProcessor<'a> {
+    definitions: &'a mut HashMap<String, (Symbol, Option<FunctionValueDefinition>)>,
+    implementations: &'a HashMap<String, Vec<TraitImplementation<Expression>>>,
+    type_args_stack: Vec<Vec<Type>>,
+    stack: Vec<String>,
+    entry_point: Option<FunctionValueDefinition>,
+}
+
+impl<'a> TraitsProcessor<'a> {
+    pub fn new(
+        definitions: &'a mut HashMap<String, (Symbol, Option<FunctionValueDefinition>)>,
+        implementations: &'a HashMap<String, Vec<TraitImplementation<Expression>>>,
+    ) -> Self {
+        Self {
+            definitions,
+            implementations,
+            type_args_stack: Vec::new(),
+            stack: Vec::new(),
+            entry_point: None,
+        }
+    }
+
+    pub fn traits_resolution(&mut self) {
+        let keys = self
+            .definitions
+            .iter()
+            .filter(|(_, (_, def))| {
+                def.is_some()
+                    && !matches!(
+                        def.as_ref().unwrap(),
+                        FunctionValueDefinition::TraitFunction(_, _)
+                    )
+            })
+            .map(|(name, _)| name.clone())
+            .collect::<Vec<_>>();
+        for name in keys {
```

Comment:
I think we should also filter out every definition that has type variables

Reply:
Shouldn't it be the other way around? Shouldn't it filter out all those that don't have type vars?